sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/AbstractEventSourcedEntityHandler.scala

```diff
@@ -0,0 +1,113 @@
+import com.akkaserverless.javasdk.impl.eventsourcedentity.EventSourcedEntityEffectImpl.EmitEvents
+import com.akkaserverless.javasdk.impl.eventsourcedentity.EventSourcedEntityEffectImpl.NoPrimaryEffect
+import com.google.protobuf.{Any => JavaPBAny}
+
+import java.util.Optional
+
+object AbstractEventSourcedEntityHandler {
+  case class CommandResult(events: Vector[Any],
+                           secondaryEffect: SecondaryEffectImpl,
+                           snapshot: Option[Any],
+                           endSequenceNumber: Long)
+}
+
+/**
+ * @tparam S the type of the managed state for the entity
+ * Not for manual user extension or interaction
+ */
+abstract class AbstractEventSourcedEntityHandler[S, E <: EventSourcedEntityBase[S]](protected val entity: E) {
+
+  // ugh, remove
+  var snapshotEvery = 0
+
+  private var state: Option[S] = None
+
+  final protected def stateOrEmpty(): S = state match {
+    case None =>
+      val emptyState = entity.emptyState()
+      require(emptyState != null, "Entity empty state is not allowed to be null")
+      state = Some(emptyState)
+      emptyState
+    case Some(state) => state
+  }
+
+  final protected def setState(newState: S): Unit =
+    state = Some(newState)
+
+  // "public" api against the impl/testkit
+  final def handleSnapshot(snapshot: S): Unit = setState(snapshot)
+  final def handleEvent(event: Object, context: EventContext): Unit = {
+    entity.setEventContext(Optional.of(context))
+    try {
+      val newState = handleEvent(stateOrEmpty(), event)
+      setState(newState)
+    } finally {
+      entity.setEventContext(Optional.empty())
+    }
+  }
+  final def handleCommand(commandName: String,
+                          command: JavaPBAny,
+                          context: CommandContext,
+                          eventContextFactory: Long => EventContext): CommandResult = {
+    val commandEffect = try {
+      entity.setCommandContext(Optional.of(context))
+      handleCommand(commandName, stateOrEmpty(), command).asInstanceOf[EventSourcedEntityEffectImpl[Any]]
+    } finally {
+      entity.setCommandContext(Optional.empty())
+    }
+    var currentSequence = context.sequenceNumber()
+    commandEffect.primaryEffect match {
+      case EmitEvents(events) =>
+        var shouldSnapshot = false
+        events.foreach { event =>
+          try {
+            entity.setEventContext(Optional.of(eventContextFactory(currentSequence)))
+            val newState = handleEvent(stateOrEmpty(), event)
+            setState(newState)
+          } finally {
+            entity.setEventContext(Optional.empty())
+          }
+          currentSequence += 1
+          shouldSnapshot = shouldSnapshot || (snapshotEvery > 0 && currentSequence % snapshotEvery == 0)
+        }
+        // FIXME currently snapshotting final state after applying all events even if trigger was mid-event stream?
```

Comment:
I think this is ok since snapshotting is just an optimization. We don't need to be precise about what was exactly the event that reached the threshold